docs(guide): added schema first guide (#5787)
mikro-orm/mikro-orm: docs/docs/schema-first-guide.md

-- Table `article_tag`
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS `article_tag` (
  `article_id` INT UNSIGNED NOT NULL,
  `tag_id` INT UNSIGNED NOT NULL,
  PRIMARY KEY (`article_id`, `tag_id`),
  INDEX `fk_article_tag_tag1_idx` (`tag_id` ASC) VISIBLE,
  CONSTRAINT `fk_article_tag_article1`
    FOREIGN KEY (`article_id`)
    REFERENCES `article` (`id`)
    ON DELETE NO ACTION
    ON UPDATE NO ACTION,
  CONSTRAINT `fk_article_tag_tag1`
    FOREIGN KEY (`tag_id`)
    REFERENCES `tag` (`id`)
    ON DELETE NO ACTION
    ON UPDATE NO ACTION)
ENGINE = InnoDB;

SET SQL_MODE=@OLD_SQL_MODE;
SET FOREIGN_KEY_CHECKS=@OLD_FOREIGN_KEY_CHECKS;
SET UNIQUE_CHECKS=@OLD_UNIQUE_CHECKS;
```

But we can place this in an initial migration file, to make our application work on blank MySQL servers as well.

## Project setup

### Install

We will use a similar setup to [the guide](./guide).

Init the project:
```bash npm2yarn
npm init
```

Install the following:
```bash npm2yarn
npm install @mikro-orm/core \
            @mikro-orm/mysql \
            @mikro-orm/migrations \
            fastify
```

and some dev dependencies

```bash npm2yarn

> npm install --save-dev @mikro-orm/cli \

Reviewer: hmm i thought `--dev` works the same, if not, we should alter the getting started guide too

also the indent is broken because of this

                  @mikro-orm/entity-generator \
                  typescript \
                  ts-node \
                  @types/node \
                  rimraf \
                  vitest
```

### ECMAScript Modules

> Just [as in the "code first" guide](./guide/01-first-entity.md#ecmascript-modules), we'll be using ECMAScript modules. Make sure you have

Reviewer: i would maybe explicitly mention that people can still use CJS, since this sounds a bit like you need to use ESM

in the long run, i guess we should also have CJS versions of both guides, or maybe just a CJS version of the final project in a branch.

```json title='package.json'
{
  "type": "module",
  ...
}
```

in your package.json file.

### Configuring TypeScript

We will use almost the same config [as the "code first" guide one](./guide/01-first-entity.md#configuring-typescript). As mentioned there, already, adjust this config if you know what you’re doing.

We'll include the ts-node config, and add `emitDecoratorMetadata`, because we'll be using the default metadata provider, which requires that of our TypeScript config.

```json title='tsconfig.json'
{
  "compilerOptions": {
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    "target": "ES2022",
    "strict": true,
    "outDir": "dist",
    "declaration": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true
  },
  "include": [
    "./src/**/*.ts"
  ],
  "ts-node": {
    "esm": true,
    "transpileOnly": true
  }
}
```

### Configuring the CLI

> Configuring the MikroORM CLI tools is essential to a "schema first" approach.  We need the migrator to run our initial and later migration, as well as the entity generator, to create our entities out of the schema state.

Reviewer: ```suggestion
Configuring the MikroORM CLI tools is essential for the "schema first" approach.  We need the migrator to run our migrations, as well as the entity generator to create our entities out of the schema state.
```
